jax/_src/lax/control_flow/loops.py

```diff
@@ -1618,6 +1674,79 @@ def _while_typecheck(_, *in_atoms, cond_jaxpr, body_jaxpr, cond_nconsts,
         f'Effects not supported in `while`: {disallowed_effects}')
   return body_jaxpr.out_avals, joined_effects
 
+def _while_discharge_rule(in_avals, out_avals, *args, cond_jaxpr, body_jaxpr,
+                          cond_nconsts, body_nconsts):
+  # TODO(sharadmv): enable supporting state effects in the cond
+  if any(isinstance(eff, state.RefEffect) for eff in cond_jaxpr.effects):
+    raise NotImplementedError
+  cond_consts, body_consts, carry = split_list(args, [cond_nconsts, body_nconsts])
+  cond_consts_avals, body_consts_avals, carry_avals = split_list(in_avals,
+                                                                 [cond_nconsts,
+                                                                  body_nconsts])
+  # There shouldn't be any `Ref`s in the `cond` (because of our check above).
+  assert not any(isinstance(aval, state.AbstractRef) for aval in cond_consts_avals)
+  is_ref = [isinstance(aval, state.AbstractRef) for aval in body_consts_avals]
+  remaining_body_consts, refs = partition_list(is_ref, body_consts)
+  remaining_body_const_avals, ref_avals = partition_list(is_ref,
+                                                         body_consts_avals)
+  num_refs = sum(is_ref)
+  num_remaining_consts = body_nconsts - num_refs
+  num_carry = len(in_avals) - body_nconsts - cond_nconsts
+  body_jaxpr, body_jaxpr_consts = body_jaxpr.jaxpr, body_jaxpr.consts
+  cond_jaxpr, cond_jaxpr_consts = cond_jaxpr.jaxpr, cond_jaxpr.consts
+  if body_jaxpr_consts: raise NotImplementedError
```

Comment:
I think I know how to hit this one! Maybe also add an "if you hit this" message.

Let's save fixing it for follow-up work but it might behoove us to fix in the near future while some of this is fresh in our minds.

Reply:
Done.